### PR TITLE
feat: makes mutual TLS optional for postgres, mysql/mariadb and grpc

### PR DESCRIPTION
**Pull request by NickUfer**
This PR makes it possible to create a secure connection to a postgresql, mariadb/mysql or grpc server without the requirement of a client certificate. Mutual TLS would be optional.

### CODE REVIEW

**Comment by iegomez** [comment]
@NickUfer thanks for this new PR!

I see you added new commits a couple of hours ago, so I believe it's still a WIP. Also, could you throw in some tests for the changes? Let me know when it's ready for review.

Cheers!

**Comment by NickUfer** [comment]
In the latest commits I added basic support for tls and mutual tls tests and in the coming days I will add the actual tests for both tls and mutual tls :+1:

**Comment by iegomez** [comment]
That's awesome, thank you very much!

**Comment by NickUfer** [comment]
> That's awesome, thank you very much!

Thank you and the other contributors for creating this project in the first place. It will be pretty valuable to me 😄


I think the PR is ready for review 👍🏼 Tests should run through. In the process I found some problems with the tls implementation in pgsql and grpc, these problems should be fixed too.

**Comment by iegomez** [comment]
@NickUfer sorry for the late reply, I've been a bit busy and missed your comment. I'll take a new look during the week.

File: backends/postgres.go [hunk 1]
```diff
@@ -133,13 +147,24 @@
 	connStr := fmt.Sprintf("user=%s password=%s dbname=%s host=%s port=%s", postgres.User, postgres.Password, postgres.DBName, postgres.Host, postgres.Port)
 
 	if (postgres.SSLMode == "verify-ca" || postgres.SSLMode == "verify-full") && checkSSL {
-		connStr = fmt.Sprintf("%s sslmode=verify-ca sslcert=%s sslkey=%s sslrootcert=%s", connStr, postgres.SSLCert, postgres.SSLKey, postgres.SSLRootCert)
-	} else if postgres.SSLMode == "require" {
+		fmt.Printf("%s sslmode=%s sslrootcert=%s\n", connStr, postgres.SSLMode, postgres.SSLRootCert)
```
[the remark this answers is not in the document]
Comment (iegomez) [reply backends/postgres.go:150]: @NickUfer as the warnings show, this leaks credentials from the connection string.

File: backends/postgres.go [hunk 2]
```diff
@@ -133,13 +147,24 @@
...
+		connStr = fmt.Sprintf("%s sslmode=%s sslrootcert=%s", connStr, postgres.SSLMode, postgres.SSLRootCert)
+	} else if postgres.SSLMode == "require" && checkSSL {
+		fmt.Printf("%s sslmode=require\n", connStr)
```
[the remark this answers is not in the document]
Comment (iegomez) [reply backends/postgres.go:153]: Same.

File: backends/postgres.go [hunk 3]
```diff
@@ -133,13 +147,24 @@
...
 		connStr = fmt.Sprintf("%s sslmode=require", connStr)
 	} else {
+		fmt.Printf("%s sslmode=disable\n", connStr)
```
[the remark this answers is not in the document]
Comment (iegomez) [reply backends/postgres.go:156]: Same.

**Review by iegomez — CHANGES_REQUESTED** [review]
There are some sensitive info leaks. Besides that, all looks good! 🚀

**Comment by iegomez** [comment]
@NickUfer the code scan caught some leaks that need to be addressed. Other than that, could you target `release-2.0.0` so @maab may then rebase their PR and merge as well? Afterwards I'll merge that into master and make a release.

**Comment by NickUfer** [comment]
@iegomez everything is ready now after some delays because I got sick

File: backends/postgres.go [hunk 4]
```diff
@@ -98,6 +98,14 @@ func NewPostgres(authOpts map[string]string, logLevel log.Level, hasher hashing.
 	}
 
 	if sslmode, ok := authOpts["pg_sslmode"]; ok {
+		switch sslmode {
+		case "verify-full":
```
Comment (iegomez) [inline backends/postgres.go:102]: Cases may be combined as comma separated values if I recall correctly.

Comment (NickUfer) [reply backends/postgres.go]: Fixed

**Comment by iegomez** [comment]
Thanks, @NickUfer!

I need to run an errand now, I'll merge and release when I get back.

**Review by iegomez — APPROVED** [review]

**Comment by maab** [comment]
Thank you very much @NickUfer and @iegomez for the work you had with reviewing and integrating this great PR!

**Pull request merged**